### ISSUE TITLE
Contention around net.osmand.plus.resources.AmenityIndexRepositoryBinary.searchAmenities

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#19925]
Upon researching further ways to speed up app start, this finding strikes me: On some occasions pretty soon after app start, there is a condition leading to the app being unresponsive for about >30 sec, and I have captured it in this [logcat_30sec-hang.log.txt](https://github.com/osmandapp/OsmAnd/files/15400380/logcat_30sec-hang.log.txt).

The app "hang" coincides with the 19 `Long monitor contention with owner` lines in the 50 sec between 05-22 09:18:14.286 and 05-22 09:19:04.890, all centering about a contention around `java.util.List net.osmand.plus.resources.AmenityIndexRepositoryBinary.searchAmenities(...)`.

It's a timing issue, and seems to primarily occur if in the first seconds after app start the user immediately starts some expensive operations, like opening MyPlaces>Tracks. (In which case `TrackFolderLoaderTask.loadGPXFolder(...)` is triggered and competes with `gpxDbHelper.loadItems()`, causing high load to the system. In these cases `GPX_DB_INITIALIZED` reports 8000ms in my setup, while if left alone during app start we have to down to 2000ms now.

Even though it is not a regular occurrence, it may be worth investigating and mitigating, it is one of the last few things I see where we could clean up our app start.

**Comment by sonora** [comment]
> * Have no idea about searchAmenities. Likely it's indexing process and trying to reindex to find nearest city for tracks

Yes, that was my first impulse, too. But if true, this seems buggy, because one of the "laggiest" processes we now have is actually going to "Search", then tapping on "Address search", and then waiting for the screen to show the "Nearest cities".

I believe there may be significant bugs why that takes so long... (or potentially using too much expensive I/O, compare to how we have lately managed to speed up MyPlaces>Tracks  :wink:).

Actually, @Chumva has a brilliant point here: https://github.com/osmandapp/OsmAnd-Issues/issues/1982

**Comment by sonora** [comment]
@vshcherb I found the culprit: If you enter the `MyPlaces>Tracks` screen too early, the condition `item.getAnalysis == null` in 
https://github.com/osmandapp/OsmAnd/blob/3ffea956e4b8e7a736e99da8e848605cac312cb5/OsmAnd/src/net/osmand/plus/track/helpers/GpxDbUtils.java#L222
causes an update storm of 
https://github.com/osmandapp/OsmAnd/blob/3ffea956e4b8e7a736e99da8e848605cac312cb5/OsmAnd/src/net/osmand/plus/track/helpers/GpxReaderTask.java#L90
putting the app in the unresponsive state.

This fixes it: https://github.com/osmandapp/OsmAnd/commit/fe1828b73770a4dfac018c0ecf17e1b766405621

We have a number of reported hangs in our Issues list, I think I remember e.g. by @Rowin63, which may also be fixed by this.

**Comment by vshcherb** [comment]
Yes that fixes analysis that is't not needed however it's not an infinite loop

```
		if (GpxDbUtils.isAnalyseNeeded(item) && !isGpxReading(file)) {
```
There is a check that file will get analysis once per start however it's another bug obviously

**Comment by sonora** [comment]
Yes, the issue I was fixing was an annoying bug early in the app lifecycle sometimes stalling the entire app with Android reporting it as non-responsive.

At that point, `getAnalysis()` still reported as "null" (but that is a transient condition at this time), and mass-triggered `updateGpxDataItem(...)`, although this is not needed to produce the Tracks screen correctly.

The other suspicion you raise is justified and deserves separate investigation. I concur that we may still have too many false `isAnalyseNeeded(item) == true's`, but with all the recent improvement, even with my extensive collection of GPX files on the device  (coming from 40 sec to open the Tracks screens to now 1-3 sec, and startup times of the app down to ~6 sec), further optimization has become harder for me to even measure a KPI for. Eventually it would make sense, and I am willing to revisit, as part of #19683.

**Comment by sonora** [comment]
_Thanks for the flowers_, @pebogufi and @Rowin63.  😉😉😉

I was amazed myself how much you can achieve if something has gotten on your nerves for long enough in your daily life so that you finally decide you invest 2 weekends or so to educate yourself, investigate some code, and optimize it.

I found the fun about optimizing code is it saves CPU cycles, power consumption, CO2, and ulitimately all of our life time not having to wait so long for an app to respond to what you want it to do ... 😉

But let me extend the Thank You to the OsmAnd dev team who have reviewed and corrected my commits, underpinning this success story.
